src/main/java/de/tum/cit/aet/artemis/exercise/service/ParticipationFilterService.java

```diff
@@ -0,0 +1,117 @@
+package de.tum.cit.aet.artemis.exercise.service;
+
+import static de.tum.cit.aet.artemis.core.config.Constants.PROFILE_CORE;
+
+import java.util.ArrayList;
+import java.util.Optional;
+import java.util.Set;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+import org.springframework.context.annotation.Profile;
+import org.springframework.stereotype.Service;
+
+import de.tum.cit.aet.artemis.assessment.domain.Result;
+import de.tum.cit.aet.artemis.exercise.domain.Exercise;
+import de.tum.cit.aet.artemis.exercise.domain.ExerciseType;
+import de.tum.cit.aet.artemis.exercise.domain.Submission;
+import de.tum.cit.aet.artemis.exercise.domain.participation.Participation;
+import de.tum.cit.aet.artemis.exercise.domain.participation.StudentParticipation;
+
+/**
+ * A service to handle participation filtering.
+ */
+@Service
+@Profile(PROFILE_CORE)
+public class ParticipationFilterService {
+
+    private final SubmissionFilterService submissionFilterService;
+
+    public ParticipationFilterService(SubmissionFilterService submissionFilterService) {
+        this.submissionFilterService = submissionFilterService;
+    }
+
+    /**
+     * Finds all participations of a student in the given exercise. For non-programming exercises, there should only be
+     * at most one participation. For programming exercises, an additional practice participation can exist.
+     *
+     * @param participationsAcrossAllExercises all participations of the student in all exercises of the course
+     * @param exercise                         the exercise for which we want the participations
+     * @return the participations in the exercise
+     */
+    public Set<StudentParticipation> findStudentParticipationsInExercise(Set<StudentParticipation> participationsAcrossAllExercises, Exercise exercise) {
+        // only consider participations in the given exercise
+        if (participationsAcrossAllExercises == null || participationsAcrossAllExercises.isEmpty()) {
+            return Set.of();
+        }
+        var participationsInExercise = participationsAcrossAllExercises.stream().filter(p -> p.getExercise() != null && p.getExercise().equals(exercise))
+                .collect(Collectors.toSet());
+
+        if (participationsInExercise.isEmpty()) {
+            return Set.of();
+        }
+
+        if (ExerciseType.PROGRAMMING.equals(exercise.getExerciseType())) {
+            return findStudentParticipationsForProgrammingExercises(participationsInExercise);
+        }
+        else {
+            return Set.of(findStudentParticipationForNonProgrammingExercises(participationsInExercise));
+        }
+    }
+
+    /**
+     * Filters the student participation for display in the student dashboard.
+     *
+     * @param participation the participation that contains the exercise and should be filtered
```

Comment:
```suggestion
     * Filters the student participation for display in the student dashboard.
     *
     * <p>Ensures that sensitive information (e.g. the assessing tutor) is removed. Additionally, this only keeps the latest submission and its most recent visible result. For example, in case the exercise uses a manual assessment and the assessment publication date has not yet passed, only the last automatic result is kept since the newer one should not yet be visible to the student.
     *
     * @param participation the participation that contains the exercise and should be filtered
```
I think the Javadoc here would benefit a bit from explaining what is retained and what is removed as part of the participation in the filtering step.

(Checkstyle/Spotless will probably complain about the line length when using the suggestion directly. The comment describes what I understood from the implementation, feel free to remove/add/adapt as you deem fit for stuff that is inaccurate.)